### DISCUSSION TITLE
Great project

### DISCUSSION THREAD

**Discussion by fritz47110815** [beep-projects/SystaPi#4] [Ideas]
Thanks a lot for this great project. I tested it successfully with a raspberry 3b plus USB wifi adapter.
The Systa has the version 1.12 20.05.14 - unfortunately I don't know how to find out System and Basis versions.

Do you think it is possible to add the modification one or more parameter (i.e. Operating Mode)? If you want I could provide some information.

**Comment by beep-projects** [comment]
Hi fritz,

for sure I can add changes, but first you have to be more specific on what your request is.
The operating mode is returned as part of 
GET /SystaREST/status http://systapi:1337/SystaREST/status

just send a note on which parameters you would like to get and I can add these.

Are the values returned by status plausible for your system? What components do you have attached to your system?

I get the other software version details if I navigate on the control unit to the menu entry where it shows the software version of the SystaComfort and then press the double down arrows on the right of the button field

**Comment by fritz47110815** [comment]
Hi,
I am sorry for the misunderstanding. What I mean is something like
http://systapi:1337/SystaREST/setparameter?OperatingMode=Automatic1
Caused by this example the operating mode of the systa should be switched. I can provide some information of the specific communication protocol to change parameters.
Thx for the instruction: System V2.09.1   / Basis V0.23

**Reply by beep-projects** [reply]
I am currently looking into setting the parameters, based on what the S-touch does (weird protocol).
So if you have any information about how to set parameters, I am happy to add this feature

**Reply by beep-projects** [reply]
btw. I am starting off from this  https://community.openhab.org/uploads/short-url/d9EE8RfvameaneEkC6oZqHlrDdM.txt but that seems to be a copy & paste of the packets sent by the s-touch and I would like to understand that protocol more in-depth before I create an API for it

**Comment by fritz47110815** [comment]
I did some investigations:

Ususally the datagrams will be initiated by the systa: it provides at least the MAC address and the counter. The portal answers with the MAC and the counter.
When a parameter change has been done at the portal it modifies its answer by changing the record type to 0x02 at datagram position 12 (which is usually 0x01). In addition the Systa password will be transferred, which is by default 1212. When Systa will receive such a datagram it will answer with some stuff (could be ignored) and increases the counter. Now the portal will send another datagram of type 0x02 including the modified parameter and new value of the parameter. Last not least Systa will send a confirmation datagram (could be ignored) and the parameter has been changed.


**Example of a datagram flow** (I masked the first 4 bytes of the MAC with aabbccdd):

normal datagram of type 2 sent by Systa
```
aa bb cc dd 0f 37 79 24 09 09 0c 00 32 dc 00 00 02 00 00 00 00 ... 00 00 00 00 00 00 aa 88 4d a2
                                                ^
                                                a
a - telegram type 255
``` 

Portal sends the password 
```
aa bb cc dd 0f 37 79 24 01 00 00 00 02 00 08 00 00 00 00 31 32 31 32 00 c4 f6 72 55  
                                     ^                    ^             ^     ^
                                     a                    b             c     d
```
a - datagram type a
b - password in ASCII
c - 1st checksum
d - 2nd checksum

1st checksum will be built by adding the last two bytes of the mac plus the specific offset 0xBFB5
2nd checksum will be build by adding the counter + 0x10F9 (or 30F9 or 50F9)

Systa answers
```
aa bb cc dd 0f 37 7a 24 09 09 0c 00 32 dc 00 00 ff ff ff ff 02 02 08 02 f2 c5 59 d1 9d 
                                                ^
                                                a
a - telegram type 255
```
 
Portal sends the parameter modification
```
aa bb cc dd 0f 37 7a 24 01 00 00 00 02 00 11 01 1c 00 00 00 01 00 8f c5 4a 65  
            ^     ^                 ^     ^  ^           ^        ^     ^
            a     b                 c     d  e           f        g     h
a - last 2 bytes of the MAC
b - 2 byte counter
c - datagram type a
d - portal announced a parameter change
e - 2 byte code of the parameter (I started with a small list)
f - 2 byte value of the parameter (as usual Big Endian)
g - first checksum
h - second checksum
```
 
Systa answers
```
aa bb cc dd 0f 37 7b 24 09 09 0c 00 32 dc 00 00 ff ff ff ff 02 04 00 11 01 1c cc 36 47 04 ac  
                                                ^
                                                a
a - telegram type 255
```

Now Portal answers with the usual handshake.


**Explanations**

**Code of the parameter**
The 2 byte code of the parameter needs to find out with try and error. Here is the list of 4 parameter:
```
01 1c  Operating Mode (Betriebsart) ... 0=Auto1 1=Auto2 ...
01 1f  Room temp heat (Raum Temp Heizen)
01 20  Room temp comf (Raum Temp Komf)
01 21  Room temp low. (Raum Temp Offset)
```


This biggest challange were the checksums, which was more than a mess. And I am still
not 100% sure if my assumptions are correct.

**first checksum**
It is based on the last two bytes of the MAC which is byte 4 and 5 of the telegram
(in my case 0f 37), a parameter specific offset and the send byte of the value.
So using this example:
0x370f + 0x8e7f + 0x01 = 0xc58f ... byte switch "8f c5"
Here are the parameter specific offsets for the table above:
```
01 1c  0x8e7f  Operating Mode (Betriebsart) ... 0=Auto1 1=Auto2 ...
01 1f  0x8e7e Room temp heat (Raum Temp Heizen)
01 20  0x8ea3 Room temp comf (Raum Temp Komf)
01 21  0x8eae Room temp low. (Raum Temp Offset)
```

**second checksum**
It is based on counter (bytes 6 and 7), Portal Announcment Byte (=0x11 ), the CounterOffset and a multiple of 256.
So using this example:
0x247a + 0x11 + 0x3fbf + 0x100 = 0x654a

I was unable to find out why sometimes 0x100 sometimes 0x200 sometimes 0x300 will be used


Feel free to ask, if I was not clear

**Reply by beep-projects** [reply]
Thank you for this description!
I don't have an account for Systa Web, so I haven't yet investigated this communication.
I'm currently investigating the communication between my SystaComfort and the S-Touch app. But the app uses a strange protocol where the SystaComfort paints elements on the canvas in the app and both systems exchange xy coordinates of touch events. I have a good understanding of this protocol, but using it for automation is difficult because you have to simulate the touch operation of the app and can't send direct commands. I'll give it a few more tries, but if I don't get anywhere with it, I'll continue based on your work

**Comment by beep-projects** [comment]
I am trying to follow your instructions, but fail with the first packet
I try to announce the parameter change with a type 2 packet and the password.
For that I have some questions.
- Is the parameter change always replied on a type 2 packet from the SysteComfort?
- The type 1 reply is 20 byte, the type 2 reply is 28 byte long?
- The last 4 byte are the checksum calculated as for the type 1 reply?
- For the type 1 reply, the first 8 byte are copied from the last message received from the SystaComfort, followed by 0 up to byte 12 (the message type byte). In your example the 9th byte is 1. Is that also a difference to the normal reply?

**Comment by fritz47110815** [comment]
..    Is the parameter change always replied on a type 2 packet from the SysteComfort?
No, I have seen a type 2 answer on any packet from the Systa
..    The type 1 reply is 20 byte, the type 2 reply is 28 byte long?
Yes, they are different, but the 1st answer containing the password has 28 and the 2nd is 20.
..    The last 4 byte are the checksum calculated as for the type 1 reply?
Well, they are two independend checksums of 2 bytes and they are calculated as described above
..    For the type 1 reply, the first 8 byte are copied from the last message received from the SystaComfort, followed by 0 up to byte 12 (the message type byte). In your example the 9th byte is 1. Is that also a difference to the normal reply?
You are right, I haven't noticed that.

I could send you another stream of bytes doing a change of the same parameter if you are interested, are you?

**Reply by beep-projects** [reply]
sure, every packet stream will help

**Reply by beep-projects** [reply]
best would be a pcap of the communication

**Comment by fritz47110815** [comment]
sent message via Facebook

**Comment by errut** [comment]
seem not to work anymore. It installs OS but i do not find SystaPi-2.0
I am on Ubuntu , is there a Chance to get an Update?
I tried it on Windows, no success .
SW Version: Software V2.16 9.03.21 and RPI3
Appriciation
